crates/bevy_asset/src/meta.rs

```diff
@@ -0,0 +1,250 @@
+            Some(settings)
+        } else {
+            None
+        }
+    }
+    fn loader_settings_mut(&mut self) -> Option<&mut dyn Settings> {
+        if let AssetAction::Load { settings, .. } = &mut self.asset {
+            Some(settings)
+        } else {
+            None
+        }
+    }
+    fn processed_info(&self) -> &Option<ProcessedInfo> {
+        &self.processed_info
+    }
+    fn processed_info_mut(&mut self) -> &mut Option<ProcessedInfo> {
+        &mut self.processed_info
+    }
+}
+
+impl_downcast!(AssetMetaDyn);
+
+/// Settings used by the asset system, such as by [`AssetLoader`], [`Process`], and [`AssetSaver`]
+///
+/// [`AssetSaver`]: crate::saver::AssetSaver
+pub trait Settings: Downcast + Send + Sync + 'static {}
+
+impl<T: 'static> Settings for T where T: Send + Sync {}
+
+impl_downcast!(Settings);
+
+/// The () processor should never be called. This implementation exists to make the meta format nicer to work with.
+impl Process for () {
+    type Settings = ();
+    type OutputLoader = ();
+
+    fn process<'a>(
+        &'a self,
+        _context: &'a mut bevy_asset::processor::ProcessContext,
+        _meta: AssetMeta<(), Self>,
+        _writer: &'a mut bevy_asset::io::Writer,
+    ) -> bevy_utils::BoxedFuture<'a, Result<(), bevy_asset::processor::ProcessError>> {
+        unreachable!()
+    }
+}
+
+impl Asset for () {}
+
+impl VisitAssetDependencies for () {
+    fn visit_dependencies(&self, _visit: &mut impl FnMut(bevy_asset::UntypedAssetId)) {
+        unreachable!()
+    }
+}
+
+/// The () loader should never be called. This implementation exists to make the meta format nicer to work with.
+impl AssetLoader for () {
+    type Asset = ();
+    type Settings = ();
+    fn load<'a>(
+        &'a self,
+        _reader: &'a mut crate::io::Reader,
+        _settings: &'a Self::Settings,
+        _load_context: &'a mut crate::LoadContext,
+    ) -> bevy_utils::BoxedFuture<'a, Result<Self::Asset, anyhow::Error>> {
+        unreachable!();
+    }
+
+    fn extensions(&self) -> &[&str] {
+        unreachable!();
+    }
+}
+
+pub(crate) fn loader_settings_meta_transform<S: Settings>(
+    settings: impl Fn(&mut S) + Send + Sync + 'static,
+) -> MetaTransform {
+    Box::new(move |meta| {
+        if let Some(loader_settings) = meta.loader_settings_mut() {
+            if let Some(loader_settings) = loader_settings.downcast_mut::<S>() {
+                settings(loader_settings);
+            } else {
+                error!(
+                    "Configured settings type {} does not match AssetLoader settings type",
+                    std::any::type_name::<S>(),
+                );
+            }
+        }
+    })
+}
+
+pub type AssetHash = [u8; 16];
+
+/// NOTE: changing the hashing logic here is a _breaking change_ that requires a [`META_FORMAT_VERSION`] bump.
+pub(crate) fn get_asset_hash(meta_bytes: &[u8], asset_bytes: &[u8]) -> AssetHash {
+    let mut context = md5::Context::new();
```

Comment:
Even without considering security, using md5 is not a good choice. siphash is usually a better choice. siphash as short hash is faster than md5, has no known weaknesses and is more secure.

rustc also uses siphash for similar purposes. https://github.com/rust-lang/rust/blob/481d45abeced571b533016a994cba7337102a4a4/compiler/rustc_data_structures/src/stable_hasher.rs#L22

Reply:
TBH the fact that not even the authors of SipHash are willing to call it cryptographically secure tells me its not.

But md5 is a known security weakness ancd must not be considered secure under FIPS.

Reply:
Yes, short hash output 64bit space, which is too small and cannot cryptographically secure. but it is not like md5 that has been broken.

Reply:
Does a hash algorithm in an asset system need to be cryptographically secure?

Reply:
> Why does a hash algorithm in an asset system need to be cryptographically secure?

For game security and anti-cheating.

The bevy engine supporting arbitrary injection of game asses is a security flaw that makes cheating in games via things like ESP easier, for example; And weak hashes on game assets allow has collisions to replace arbitrary game assets to make that easier.

Reply:
md5 has been replaced by BLAKE3 https://en.wikipedia.org/wiki/BLAKE_(hash_function)#BLAKE3